From aacf3991f1b58ccdd51fc8626c48fe47ded1f96a Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Wed, 13 Sep 2023 13:44:00 +0200
Subject: [PATCH 340/841] feat(stm32mp2): keep RETRAM in Standby mode

To keep RETRAM in Standby mode, the register PWR_CR10 should be set to 1,
to keep the its regulator on. And its content should not be erased,
by setting SRAMHWERDIS in RAMCFG_RETRAMCR register.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 4a9bedee024aaff097f1b27fece344c130584c68 feat(stm32mp2): keep RETRAM in Standby mode]
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Change-Id: I56b7836b51a78cfefbb9cb124c134fea28dea481
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/327287
Tested-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 include/drivers/st/stm32mp2_pwr.h |  3 +++
 plat/st/stm32mp2/stm32mp2_def.h   |  1 +
 plat/st/stm32mp2/stm32mp2_pm.c    | 10 ++++++++++
 3 files changed, 14 insertions(+)

diff --git a/include/drivers/st/stm32mp2_pwr.h b/include/drivers/st/stm32mp2_pwr.h
index 9cf54518a..6db627f53 100644
--- a/include/drivers/st/stm32mp2_pwr.h
+++ b/include/drivers/st/stm32mp2_pwr.h
@@ -132,6 +132,9 @@
 /* PWR_CR10 register fields */
 #define PWR_CR10_RETRBSEN_MASK			GENMASK_32(1, 0)
 #define PWR_CR10_RETRBSEN_SHIFT			0
+#define PWR_CR10_RETRBSEN_DISABLE		0U
+#define PWR_CR10_RETRBSEN_STANDBY_VBAT		1U
+#define PWR_CR10_RETRBSEN_STANDBY		2U
 
 /* PWR_CR11 register fields */
 #define PWR_CR11_DDRRETDIS			BIT(0)
diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index 8ca0ecbdf..24ebac785 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -438,6 +438,7 @@ static inline uintptr_t tamp_bkpr(uint32_t idx)
 #define BSEC_BASE				U(0x44000000)
 #define DBGMCU_BASE				U(0x4A010000)
 #define HASH_BASE				U(0x42010000)
+#define RAMCFG_BASE				U(0x42070000)
 #define RTC_BASE				U(0x46000000)
 #define STGEN_BASE				U(0x48080000)
 #define SYSCFG_BASE				U(0x44230000)
diff --git a/plat/st/stm32mp2/stm32mp2_pm.c b/plat/st/stm32mp2/stm32mp2_pm.c
index b8b1c53fa..8d29e61c9 100644
--- a/plat/st/stm32mp2/stm32mp2_pm.c
+++ b/plat/st/stm32mp2/stm32mp2_pm.c
@@ -30,6 +30,9 @@
 
 #define CA35SS_SYSCFG_VBAR_CR	0x2084U
 
+#define RAMCFG_RETRAMCR		0x180U
+#define SRAMHWERDIS		BIT(12)
+
 /* GIC interrupt number */
 #define RCC_WAKEUP_IRQn		254
 
@@ -858,6 +861,13 @@ static void stm32_pm_init(void)
 
 	mmio_write_32(rcc_base + RCC_C1SREQCLRR, RCC_C1SREQSETR_STPREQ_MASK);
 
+	/* Maintain BKPSRAM & RETRAM content in Standby */
+	mmio_write_32(pwr_base + PWR_CR9, PWR_CR9_BKPRBSEN);
+	mmio_write_32(pwr_base + PWR_CR10, PWR_CR10_RETRBSEN_STANDBY);
+
+	/* Prevent RETRAM erase */
+	mmio_write_32(RAMCFG_BASE + RAMCFG_RETRAMCR, SRAMHWERDIS);
+
 	/*
 	 * If PWR_D1CR_LPCFG_D1 is set to 0,
 	 *      PWR_CPU_ON signal is low when CPU1 is in standby or stop2/stop3
-- 
2.47.2

